django/utils/functional.py

```diff
@@ -18,13 +21,30 @@ class cached_property:
     Decorator that converts a method with a single self argument into a
     property cached on the instance.
 
-    Optional ``name`` argument allows you to make cached properties of other
-    methods. (e.g.  url = cached_property(get_absolute_url, name='url') )
+    It is possible to make a cached property out of an existing method:
+    (e.g.  ``url = cached_property(get_absolute_url)`` ).
+    On Python versions before 3.6, you'll need to provide the optional ``name``
+    argument. (e.g.  ``url = cached_property(get_absolute_url, name='url')`` ).
     """
     def __init__(self, func, name=None):
+        func_name = func.__name__
+        unsupported_mangled_method = (
+            not PY36 and not name and
+            func_name.startswith('__') and not func_name.endswith('__')
```

Comment:
Might make sense to check explicitly set name too, because '__name' obviously will not work.

Reply:
?

Reply:
```
class A:
    __print = cached_property(print, '__print')
```
This will not work and we can easily detect it too.

Reply:
What should the condition be?

Reply:
I'll push my edits tomorrow.